File: go/src/reveal_cards_in_increasing_order/950.reveal_cards_in_increasing_order.go
// 2018-12-03 10:05
package main

import (
	"fmt"
	"sort"
)

type Queue []int

func (this *Queue) Poll() int {
	tmp := (*this)[0]
	*this = (*this)[1:]
	return tmp
}

func (this *Queue) Offer(x int) {
	*this = append(*this, x)
}

func (this *Queue) Empty() bool {
	return len(*this) == 0
}

func (this *Queue) Peek() int {
	return (*this)[0]
}

func deckRevealedIncreasing(deck []int) []int {
	sort.Ints(deck)
	N := len(deck)
	ret := make([]int, N)
	queue := Queue{}
	for i := 0; i < N; i++ {
		queue.Offer(i)
	}

	for i := 0; i < N; i++ {
		ret[queue.Poll()] = deck[i]
		if !queue.Empty() {
			queue.Offer(queue.Poll())
		}
	}

	return ret
}

func main() {
	// 2 3 5 7 11 13 17
	// 2 3 5 7 11 13 17 3 7 13 17 3
	fmt.Println(deckRevealedIncreasing([]int{17, 13, 11, 2, 3, 5, 7}))
}

File: rust/reveal_cards_in_increasing_order/950_reveal_cards_in_increasing_order.rs
// 2018-12-24 17:18
// @lc id=950 lang=rust
use std::collections::VecDeque;
struct Solution {}
impl Solution {
    pub fn deck_revealed_increasing(deck: Vec<i32>) -> Vec<i32> {
        let mut ret = vec![0; deck.len()];
        let mut deck = deck;
        deck.sort();
        // 2,3,5,7,11,13,17
        let mut q = VecDeque::new();
        for i in 0..deck.len() {
            q.push_back(i)
        }
        let mut index = 0;
        loop {
            if q.is_empty() {
                break;
            }
            let top = q.pop_front().unwrap();
            ret[top] = deck[index];
            index += 1;

            if q.is_empty() {
                break;
            }
            let top = q.pop_front().unwrap();
            q.push_back(top);
        }
        return ret;
    }
}

fn main() {
    println!(
        "{:?}",
        Solution::deck_revealed_increasing(vec![17, 13, 11, 2, 3, 5, 7])
    );
}
